include/range/v3/view/set_algorithm.hpp

```diff
@@ -0,0 +1,904 @@
+
+                        return state_t::FIRST;
+                    }
+                    
+                    return state_t::ONLY_SECOND;
+                }
+                
+            public:
+                using value_type = common_type_t<range_value_t<R1>, range_value_t<R2>>;
+                using reference_type = common_reference_t<range_reference_t<R1>, range_reference_t<R2>>;
+                using rvalue_reference_type = common_reference_t<range_rvalue_reference_t<R1>,
+                                                                 range_rvalue_reference_t<R2>>;
+                using single_pass = meta::strict_or<SinglePass<range_iterator_t<R1>>,
+                                                    SinglePass<range_iterator_t<R2>>>;
+
+                set_union_cursor() = default;
+                set_union_cursor(pred_ref_ pred, proj1_ref_ proj1, proj2_ref_ proj2,
+                                 range_iterator_t<R1> it1, range_sentinel_t<R1> end1,
+                                 range_iterator_t<R2> it2, range_sentinel_t<R2> end2)
+                  : pred_(std::move(pred)), proj1_(std::move(proj1)), proj2_(std::move(proj2)),
+                    it1_(std::move(it1)), end1_(std::move(end1)), it2_(std::move(it2)), end2_(std::move(end2)),
+                    state(which_set())
+                {}
+                reference_type get() const
+                noexcept(noexcept(*it1_) && noexcept(*it2_))
+                {
+                    if(state == state_t::SECOND || state == state_t::ONLY_SECOND)
+                        return *it2_;
+                    else
+                        return *it1_;
+                }
+                void next()
+                {
+                    switch(state)
+                    {
+                        case state_t::FIRST:
+                            ++it1_;
+                            state = which_set();
+                            break;
+                            
+                        case state_t::FIRST_INC_SECOND:
+                            ++it2_;
+                            ++it1_;
+                            state = which_set();
+                            break;
+                        
+                        case state_t::ONLY_FIRST:
+                            ++it1_;
+                            break;
+                            
+                        case state_t::SECOND:
+                            ++it2_;
+                            state = which_set();
+                            break;
+                            
+                        case state_t::ONLY_SECOND:
+                            ++it2_;
+                            break;
+                    }
+                }
+                bool equal(set_union_cursor const &that) const
+                {
+                    return (it1_ == that.it1_ && it2_ == that.it2_) ||
+                           (done() && that.done());
```

Comment:
> Checking `done` is extraneous here unless you want to support comparison of past-the-end iterators of differing sequences.
Exactly. I put it here in case `that` points into a subrange of one of the input ranges of `this`. Suppose `that` is already done but `this` is still not. `this` and `that` are not equal, but only comparing their member iterators could in principle say otherwise. Checking `done` covers this case.